test/src/test/java/org/corfudb/integration/CorfuReplicationMultiSourceSinkIT.java

```diff
@@ -500,18 +491,40 @@ public ReplicatedStreamsListener(CountDownLatch countdownLatch, boolean snapshot
 
         @Override
         public void onNext(CorfuStreamEntries results) {
+            // TODO: LR currently has a limitation that the Snapshot Writers on Sink cannot differentiate between
+            //  the replicated streams received from their own vs other sessions.  So if multiple writers find a
+            //  stream in the global list of streams to be replicated, each will apply updates from the shadow
+            //  stream of that stream, even though no data was received for it from the writer's own session.
+            //  In this test, Source-1 has data in Table-1, Source-2 in Table-2 and so on.  Each Source has a
+            //  separate session with the Sink.  Even though the session corresponding to Source-2 does not receive
+            //  data from Source-1(Table-1), the Sink writer will apply it because both Tables 1 and 2 are included
+            //  in the global streams to replicate set.
+            //  So the number of updates received on the Sink on a snapshot sync can be more than expected.
+            //  Hence, we perform the following modified check:
+            //  If Snapshot Sync:  Only consider the 1st set of updates(1st transaction) received for a given table.
+            //  However, this will lead to another test issue that subsequent async updates on the table will still
+            //  be received by this listener and interfere with the verification of log entry sync updates.
+            //  So we filter out these updates by looking at the number of entries in the transaction(1 entry = log
+            //  entry sync.  This assumption can work for the test as it writes single entries during log entry sync.)
+            //  Once the above limitation is addressed, these special checks must be removed.
             results.getEntries().forEach((schema, entries) -> {
-                if (snapshotSync) {
-                    if (entries.size() > 1) {
-                        processUpdate(schema, entries);
-                    }
-                } else {
+                // Ignore single 'clear' updates made to clear local writes on the Sink
+                if (snapshotSync && entries.size() > 1) {
+                    processUpdate(schema, entries);
+                } else if (!snapshotSync && entries.size() == 1) {
+                    // If in Log Entry sync, ignore updates with >1 entries (as they are from the redundant updates
```

Comment:
Not sure I understand this, why would we receive redundant updates during log entry sync? If I understand the problem occurs during snapshot sync only and basically because we apply all shadow streams. But In log entry we should differentiate updates per session, isn't it?

Reply:
Yes, we differentiate entries per session in log entry sync.

The problem is that the extra/redundant snapshot sync updates can come asynchronously after snapshot sync validation is over.  And we do not have a definite value of the countdown latch on which we can wait to be sure that all redundant updates have been applied.

Let me know if you need more explanation.